File: Algo/Algorithm/Questions/Arrays/16.subarraySort/main.js
// O(n) time | O(1) space - where n is the length of the input array
function subarraySort(array) {
  let minBreakingPoint = Infinity;
  let maxBreakingPoint = -Infinity;

  for (let i = 0; i < array.length; i += 1) {
    const breakingPoint = array[i] > array[i + 1];

    if (breakingPoint) {
      maxBreakingPoint = Math.max(maxBreakingPoint, array[i]);
      minBreakingPoint = Math.min(array[i + 1], minBreakingPoint);
    }
  }

  if (minBreakingPoint === Infinity) {
    return [-1, -1];
  }

  let leftIndex = 0;
  while (minBreakingPoint >= array[leftIndex]) {
    leftIndex += 1;
  }

  let rightIndex = array.length - 1;
  while (maxBreakingPoint < array[rightIndex]) {
    rightIndex -= 1;
  }
  return [leftIndex, rightIndex];
}

File: Algo/Algorithm/Questions/Arrays/16.subarraySort/main.cpp
#include <vector>
#include <iostream>
#include <climits>
using namespace std;

vector<int> getIndxOfSubArrayToBeSorted(int left, int right, vector<int> array)
{
  int leftIdx = 0;
  while (left >= array[leftIdx])
    leftIdx++;

  int rightIndx = array.size() - 1;
  while (right <= array[rightIndx])
    rightIndx--;

  return {leftIdx, rightIndx};
}

// O(n) time | O(1) space - where n is the length of the input array
vector<int> subarraySort(vector<int> array)
{
  int left = INT_MAX;
  int right = INT_MIN;
  for (int i = 1; i < array.size(); i++)
  {
    bool breaker = (array[i] < array[i - 1]);
    if (breaker)
    {
      left = min(array[i], left);
      right = max(array[i - 1], right);
    }
  }

  if (left == INT_MAX)
    return {-1, -1};

  return getIndxOfSubArrayToBeSorted(left, right, array);
}

int main()
{
  vector<int> array = {1, 2, 4, 7, 10, 11, 7, 12, 6, 7, 16, 18, 19};
  for (int i : subarraySort(array))
  {
    cout << i << " ";
  }
  cout << endl;
  return 0;
}
